src/systems/InternalSystem.ts

```diff
@@ -0,0 +1,57 @@
+import * as PIXI from 'pixi.js'
+import { System, Entity } from 'src/core'
+import { ShapeComponent, ComponentType } from 'src/components'
+
+export abstract class InternalSystem extends System {
+  protected app: PIXI.Application
+  protected trackedEntities: Record<string, Entity> = {}
+
+  constructor(app: PIXI.Application) {
+    super()
+    this.app = app
+  }
+
+  systemDidMount() {
+    this.engine.on('entity_added', this.addEntity)
+    this.engine.on('entity_removed', this.removeEntity)
+  }
+
+  protected addEntity = (entity: Entity) => {
+    if (this.trackedEntities[entity.uuid]) return
+
+    const shape = entity.getComponent<ShapeComponent<any>>('shape')
+
+    // Only add qualified entities
+    if (shape) {
+      if (!this.app.stage.getChildByName(entity.uuid)) {
+        const internal = new PIXI.Container()
+        internal.name = entity.uuid
+        this.app.stage.addChild(internal)
+      }
+
+      this.trackedEntities[entity.uuid] = entity
+    }
+
+    // Even if they have no Shape component now, they may have it in the future
+    entity.on('component_added', this.handleComponentAdded)
+    entity.on('component_removed', this.handleComponentRemoved)
```

Comment:
Wouldn't this fire multiple times if we add multiple entities?

Reply:
It's per component per entity. It's only fired when a component is added to *any* entity (we are tracking all of them because this is an internal system)